### PR TITLE
[GPU] Permute f and y axes

### PR DESCRIPTION
**Pull request by OlehKravchyshyn**
### Details:
 - *added kernel to switch f and y axes in 4d model blocked and planar format*
 - *added perf and unit tests*

### Ticket:
 - 92268

### CODE REVIEW

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+}
+
+static inline size_t GetTileWidth(const permute_params& params) {
+    const Datatype input_type = params.inputs[0].GetDType();
+    const Datatype output_type = params.outputs[0].GetDType();
+
+    size_t min_devisor = GetTileHeight(params);
+    // i64 only supports tile size 4
+    if ((input_type == Datatype::INT64) || (output_type == Datatype::INT64)) {
+        min_devisor = min_devisor/2;
```
Comment (tgubanova-lohika) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:65]: divisor?

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+#include <string>
+#include <functional>
+#include <cmath>
+
+namespace kernel_selector {
+
+static size_t GetDivisor(const size_t input_size) {
+    std::vector<size_t> v = {  /*32,*/ 16, 8, 4, /*2,*/ 1 };
+    auto is_devided = [input_size](int i){ return input_size%i == 0; };
+    auto result = std::find_if(begin(v), end(v), is_devided);
```
Comment (ytorzuk-altran) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:17]: What value will take tile size if input width (height) is not evenly divisible by higher then 1 values?

**Comment by ytorzuk-altran** [comment]
The idea of tiling is to avoid unordered (non-sequential) RAM transactions. Drawbacks of unordered reads from the local memory are small, therefore typical scenario is as follows:
* read from RAM sequentially
* Write to the local memory in non-sequential manner (after this stage tile should be transposed)
* Copy tile from the  local memory to RAM in sequential order

In case of bfyx->byfx permutation 3 dimensional tiles should be used to achieve performance improvements.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp (hunk 2 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:17]: for now, it is supported only cases where it is divisible by 4, 8 etc

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+        FUSED_OPS;
+        OUT_VEC_TYPE result  = FUSED_OPS_RESULT;
+#else
+        OUT_VEC_TYPE result = ACTIVATION(res, ACTIVATION_PARAMS);
+#endif
+        const int output_idx = OUTPUT_GET_INDEX(b_idx, y_idx, f_idx, x_idx);
+        WRITE_VEC(result, 0, &output[output_idx]);
+    }
+}
+#elif defined (TREE_DIM_TRANSPOSE)
```
Comment (OlehKravchyshyn) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:48]: typo -> three dim transpose

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+KERNEL (permute_f_y_axes)(
+    const __global INPUT0_TYPE* input,
+    __global OUTPUT_TYPE* output
+// TODO decide what to do with that stuff
```
Comment (OlehKravchyshyn) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:53]: remove

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+        const int output_idx = OUTPUT_GET_INDEX(b_idx, y_idx, f_idx, x_idx);
+        output[output_idx] = transpose_buf[bf_local][y_local][j];
+    }
+}
+
+#else
+KERNEL (permute_f_y_axes)(
+    const __global INPUT0_TYPE* input,
+    __global OUTPUT_TYPE* output
+// TODO decide what to do with that stuff
```
Comment (OlehKravchyshyn) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:95]: remove

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+        dispatchData.lws = GetOptimalLocalWorkGroupSizes(dispatchData.gws, params.engineInfo, in_layout, out_layout, dims_by_gws);
+    } else if (Is3DTranspose(params)) {
+        dispatchData.lws = {1, GetTileHeight(params), GetTileHeight(params)};
+    } else {
+        dispatchData.lws = {1, 1, GetTileHeight(params)};
+    }
+    return dispatchData;
+}
+
+// Validate is the same as permute_kernel_tile_8x8_4x4
```
Comment (OlehKravchyshyn) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:160]: remove

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.h [hunk 7]
```diff
@@ -0,0 +1,35 @@
+// Copyright (C) 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+//
+
+#pragma once
+
+#include "permute_kernel_base.h"
+
+namespace kernel_selector {
+////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
```
Comment (OlehKravchyshyn) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.h:10]: remove

**Comment by wilson-seok** [comment]
@OlehKravchyshyn We shared Adobe model to check performance of this PR with @mhnap-lohika. Could you please share performance on the model before/after PR?

**Comment by mhnap-lohika** [comment]
@wilson-seok We have tested this PR on the model that you shared. There is performance gain for all Permute layers, but it depends on the GPU model. We have tested only plain 4d format, as we cannot test on dGPU for now. This is the result for one of the longest-run layers "Transpose_127":

- For Intel(R) UHD Graphics 750 [0x4c8a] (iGPU) performance gain is 7.1X
- For Intel(R) Iris(R) Xe Graphics [0x9a49] (iGPU) performance gain is 3.0X

Also, I attached formatted execution graphs before and after the PR for Intel(R) Iris(R) Xe Graphics [0x9a49] (iGPU).
[permute_test_before.txt](https://github.com/openvinotoolkit/openvino/files/9949124/permute_test_before.txt)
[permute_test_after.txt](https://github.com/openvinotoolkit/openvino/files/9949125/permute_test_after.txt)

**Comment by mhnap-lohika** [comment]
> @OlehKravchyshyn We shared Adobe model to check performance of this PR with @mhnap-lohika. Could you please share performance on the model before/after PR?

Sent benchmark results in the mail.

**Comment by wilson-seok** [comment]
@OlehKravchyshyn There is performance drop in fp32/16 models. I will share the model list.

**Comment by wilson-seok** [comment]
No performance regression in fp32/16 models.

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Please resolve open conversations.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl (hunk 3 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:49]: done

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl (hunk 4 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:54]: done

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp (hunk 6 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:168]: done

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.h (hunk 7 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.h:10]: done

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl (hunk 5 above)
Comment (OlehKravchyshyn) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:105]: done

**Comment by wilson-seok** [comment]
@OlehKravchyshyn I have the performance result you provided on "Transpose_127" of target model. Those number was measured in planar format.
•	For Intel(R) UHD Graphics 750 [0x4c8a] (iGPU) performance gain is 7.1X
•	For Intel(R) Iris(R) Xe Graphics [0x9a49] (iGPU) performance gain is 3.0X
Could you please provide performance number of blocked format from unit test? We understand you guys have only iGPU and hard to measure the blocked format performance from the target model.

**Comment by wilson-seok** [comment]
@OlehKravchyshyn Please rebase this PR to resolve cldnn test failure.

**Comment by mhnap-lohika** [comment]
> @OlehKravchyshyn I have the performance result you provided on "Transpose_127" of target model. Those number was measured in planar format. • For Intel(R) UHD Graphics 750 [0x4c8a] (iGPU) performance gain is 7.1X • For Intel(R) Iris(R) Xe Graphics [0x9a49] (iGPU) performance gain is 3.0X Could you please provide performance number of blocked format from unit test? We understand you guys have only iGPU and hard to measure the blocked format performance from the target model.

@wilson-seok These are performance results from unit tests.

input: [1, 512, 16384, 1]
format: bfyx
```
Exectued time 10494.90 permute_ref
Exectued time 1132.42 permute_f_y_axes
```
input: [1, 512, 16384, 1]
format: b_fs_yx_fsv16
```
Exectued time 2779.68 permute_ref
Exectued time 1172.02 permute_f_y_axes
```
input: [1, 512, 16384, 1]
format: b_fs_yx_fsv32
```
Exectued time 2110.71 permute_ref
Exectued time 1154.89 permute_f_y_axes
```

input: [1, 256, 128, 256]
format: bfyx
```
Exectued time 1125.30 permute_ref
Exectued time 1118.79 permute_f_y_axes
```
input: [1, 256, 128, 256]
format: b_fs_yx_fsv16
```
Exectued time 4065.73 permute_ref
Exectued time 1124.21 permute_f_y_axes
```
input: [1, 256, 128, 256]
format: b_fs_yx_fsv32
```
Exectued time 3720.78 permute_ref
Exectued time 1135.75 permute_f_y_axes
```

**Comment by wilson-seok** [comment]
Please rebase this PR.

**Review by wilson-seok — APPROVED** [review]

**Comment by wilson-seok** [comment]
Please resolve the conflicts.

**Comment by wilson-seok** [comment]
@mhnap-lohika The commit you rebased seems to have build error(in transformation test of if unrolling). I rebased it.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+#else
+        OUT_VEC_TYPE result = ACTIVATION(res, ACTIVATION_PARAMS);
+#endif
+        const int output_idx = OUTPUT_GET_INDEX(b_idx, y_idx, f_idx, x_idx);
+        WRITE_VEC(result, 0, &output[output_idx]);
+    }
+}
+
+#elif defined (THREE_DIM_TRANSPOSE)
+
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:48]: I am curious why not using subgroup read/ write as my suggestion before.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+            const int x_idx = x_begin + i;
+            const int f = f_begin + j;
+            const int y_idx = y_begin + bf_local;
+            const int output_idx = OUTPUT_GET_INDEX(b_idx, y_idx, f, x_idx);
+            output[output_idx] = transpose_buf[j][bf_local][i];
+        }
+    }
+}
+
+#else
```
[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:99]: Will fix this, by the way, do you mean to use vload/vstore or BLOCK_READ macro that use subgroup extension?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl (hunk 8 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:48]: Will fix this, by the way, do you mean to use vload/vstore or BLOCK_READ macro that use subgroup extension?

**Comment by mhnap-lohika** [comment]
I did a minor refactoring but didn't change the logic. Diff can be viewed [here](https://github.com/openvinotoolkit/openvino/compare/df3ee84892..351e044dc8c2e4ce1247080ec82a06c07b58de37?diff=split).

File: src/plugins/intel_gpu/tests/test_cases/permute_gpu_test.cpp [hunk 10]
```diff
@@ -1900,3 +1963,189 @@ TEST(permute_gpu_f32_dynamic, bfyx_0_2_3_1) {
...
+        for (int i = 0; i < 10; ++i) {
+            output_permute_ref = network_ref.execute();
+            output_permute_opt = network_tile.execute();
+        }
+
+        auto r = 100;
+        double exectime_ref = 0.f;
+        double exectime_opt = 0.f;
+        for (int i = 0; i < r; ++i) {
+            //set_random_values<type>(input_opt);
```
[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/tests/test_cases/permute_gpu_test.cpp]: Done.

File: src/plugins/intel_gpu/tests/test_cases/permute_gpu_test.cpp [hunk 11]
```diff
@@ -1900,3 +1963,189 @@ TEST(permute_gpu_f32_dynamic, bfyx_0_2_3_1) {
...
     }
 }
+
+class permute_f_y_axes_tile: public TiledPermuteTest {};
+
+INSTANTIATE_TEST_SUITE_P(, permute_f_y_axes_tile,
+    ::testing::ValuesIn(std::vector<TiledPermuteParam> {
+        {{1, 4, 8, 1}, format::bfyx},
+        {{1, 4, 8, 1}, format::b_fs_yx_fsv16},
+//        {{1, 512, 16384, 1}, format::bfyx},
```
[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/tests/test_cases/permute_gpu_test.cpp]: Done.

File: src/plugins/intel_gpu/tests/test_cases/permute_gpu_test.cpp [hunk 12]
```diff
@@ -1650,31 +1709,32 @@ void TiledPermuteTest::set_random_values<int8_t>(const cldnn::memory::ptr mem) c
...
+        order.push_back(1);
+    } else {
+        std::swap(order, permute_order);
     }
-    order.push_back(1);
 
     auto input = engine.allocate_memory({Data_Type, format, tensor});
-    set_random_values<type>(input);
+    set_values<type>(input);
+    //set_random_values<type>(input);
```
[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/tests/test_cases/permute_gpu_test.cpp]: Done.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+    k.EnableBatching();
+    return k;
+}
+
+JitConstants PermuteKernel_f_y_axes::GetJitConstants(const permute_params& params,
+                                                     const CommonDispatchData& dispatchData) const {
+    auto jit = Parent::GetJitConstants(params, dispatchData);
+
+    if (params.inputs[0].X().v == 1) {
+        jit.AddConstant(MakeJitConstant("Y_WIDE_AXES", ""));
```
[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp]: No, removed.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+    } else {
+        if (IsSimpleMemCopyOperation(params)) {
+            jit.AddConstant(MakeJitConstant("PERMUTE_SIMPLE_MEM_COPY", ""));
+        }
+        if (Is3DTranspose(params)) {
+            jit.AddConstant(MakeJitConstant("THREE_DIM_TRANSPOSE", ""));
+        }
+        jit.AddConstant(MakeJitConstant("X_WIDE_AXES", ""));
```
[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp]: No, removed.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 15]
```diff
@@ -0,0 +1,227 @@
...
+}
+
+bool PermuteKernel_f_y_axes::Validate(const Params& p, const optional_params& o) const {
+    if (!Parent::Validate(p, o)) {
+        return false;
+    }
+
+    std::function<bool(const std::vector<uint16_t>&)> is_rotating_except_batch =
+        [](const std::vector<uint16_t>& order) {
+            // Target transform: Rotate feature dim to back to be taken as inner-most axis
```
[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp]: As I understood, this comment about cldnn format: bfxy with {0, 3, 2, 1} order => byxf. But in unit test, we have correct {0, 2, 1, 3} order as for IE format bfyx => byfx. We have changes in [this PR](https://github.com/openvinotoolkit/openvino/pull/11379) that convert IE permute order to cldnn permute order. But please correct me if I misunderstand something.

[the remark this answers is not in the document]
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp]: Also I have updated comment to be more clear.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl (hunk 8 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:48]: Current impl is not I suggested. I suggested to use "subgoup read" https://registry.khronos.org/OpenCL/extensions/intel/cl_intel_subgroups.html
so that each work items will read /write contiguous block in the block direction. 
Please contact me if you need further explanation. My initial suggestion for blocked format is quite simple to implement.

**Comment by akladiev** [comment]
This PR will be closed in 2 weeks in case of no activity.

**Comment by mhnap-lohika** [comment]
@yeonbok I have added the subgroup read/write to this PR, could you review it, please?

**Comment by wilson-seok** [comment]
@mhnap-lohika There is still build error in openvino-win.
image

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 16]
```diff
@@ -0,0 +1,234 @@
...
+    auto result = std::find_if(begin(v), end(v), is_divided);
+    return *result;
+}
+
+bool IsSimpleMemCopyOperation(const permute_params& params) {
+    return params.inputs[0].X().v > 1 && params.inputs[0].GetLayout() == DataLayout::bfyx;
+}
+
+bool Is3DTranspose(const permute_params& params) {
+    return params.inputs[0].X().v > 1 && params.inputs[0].GetLayout() != DataLayout::bfyx;
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:35]: I cannot understand the underlying idea for these. 
1) What is the scope of this Is3DTranspose ? All cases with X>1 && not bfyx ? Do we need this condition? can't we just do !IsSimpleMemCopyOperation() 
2) Why it is "3D" ?

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 17]
```diff
@@ -0,0 +1,234 @@
...
+}
+
+size_t GetFeatureBlockSize(const permute_params& params) {
+    return std::min(GetDivisor(params.inputs[0].Feature().v), GetDivisor(params.inputs[0].Y().v));
+}
+
+size_t GetTileHeight(const permute_params& params) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:42]: Do we need to maintain tile width/height separately? I think it make things too compilcated. Previously when we implemented permute_kernel_tile_4x4_8x8, we diddnt see that much gain with independent tile width/height. Things had better being simplified unless it is proven by an extensive perf evaluation.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp [hunk 18]
```diff
@@ -0,0 +1,234 @@
...
+    }
+    return dispatchData;
+}
+
+bool PermuteKernel_f_y_axes::Validate(const Params& p, const optional_params& o) const {
+    if (!Parent::Validate(p, o)) {
+        return false;
+    }
+
+    const auto is_rotating_except_batch = [](const std::vector<uint16_t>& order) {
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:198]: Please change this  function name, originally the "rotating" intended such transpose like [bfyx=>byxf, bfzyx => bzyxf]
where the dimensions are "rotating" except "batch" dim. 
However the target pattern of this kernel does not "rotating", it is rather, swapping ?

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl [hunk 19]
```diff
@@ -0,0 +1,195 @@
...
+    const int y_begin = get_global_id(1) * TILE_SIZE;
+    const int f_begin = get_local_size(2) * get_group_id(2);
+
+#if INPUT0_SIMPLE == 1
+    const int f = bf % INPUT0_FEATURE_NUM;
+    __attribute__((opencl_unroll_hint(TILE_SIZE)))
+    for (int j = 0; j < TILE_SIZE; ++j) {
+        const int f_idx = f_begin + j;
+        const int y_idx = y_begin + bf_local;
+        INPUT0_TYPE res = input[INPUT0_GET_INDEX(b_idx, f_idx, y_idx, x_idx)];
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:145]: Seems that this is not vectorized yet.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp (hunk 16 above)
Comment (yeonbok) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp:35]: Also if this "3D" is used for optimizing blocked format w/ sub group read/right, shouldn't we check the alignment?

File: src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp [hunk 20]
```diff
@@ -75,6 +92,52 @@ TEST(permute_gpu_f32, output_ordering_test)
...
 
+TEST(permute_f_and_y_gpu_f32, output_ordering_test)
+{
+    auto& engine = get_test_engine();
+
+    std::vector<std::vector<int32_t>> input_tensors = {
+        { 1, 256, 1024, 1 },
+    };
+    std::vector<std::vector<uint16_t>> permutations = {
+        { 0, 2, 1, 3 }, //do nothing
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp:103]: what does this "do  nothing" comment?
Also this test looks like just checking the shape... is it meaningful test?

File: src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp [hunk 21]
```diff
@@ -1677,7 +1757,7 @@ void TiledPermuteTest::run_test(const std::vector<cldnn::tensor::value_type>& si
     }
 
     auto input = engine.allocate_memory({Data_Type, format, tensor});
-    set_random_values<type>(input);
+    set_values<type>(input);
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp:1760]: why did you change this ? It is setting input only filled with "1"s...which is not good for test.

File: src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp [hunk 22]
```diff
@@ -1977,3 +2057,183 @@ TEST_P(permute_tile_fsv_5d, i64_cached) {
...
+INSTANTIATE_TEST_SUITE_P(smoke_permute_f_y_axes_tile,
+                         permute_f_y_axes_tile,
+                         ::testing::ValuesIn(std::vector<TiledPermuteParam>{
+                             {{1, 4, 8, 1}, format::bfyx},                // permute_f_y_axes
+                             {{1, 4, 8, 1}, format::b_fs_yx_fsv16},       // permute_f_y_axes
+                             {{1, 32, 256, 512}, format::b_fs_yx_fsv32},  // THREE_DIM_TRANSPOSE with subgroup
+                             {{1, 32, 256, 512}, format::bfyx},           // PERMUTE_SIMPLE_MEM_COPY
+                             {{1, 256, 256, 1}, format::b_fs_yx_fsv32},   // permute_f_y_axes
+                             {{1, 8, 4, 4}, format::b_fs_yx_fsv4},        // THREE_DIM_TRANSPOSE
+                             {{1, 4, 4, 4}, format::b_fs_yx_fsv4},        // THREE_DIM_TRANSPOSE
```
Comment (yeonbok) [inline src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp:2072]: Please add test with shapes which is not aligned with the block size

Comment (yeonbok) [reply src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp:2072]: Also can we ensure that these are selecting the new kernel? We need to enforce the implementation

**Comment by wilson-seok** [comment]
@mhnap-lohika Have you resolved all comments from Taylor?

File: src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp (hunk 21 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp]: Reverted.

File: src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp (hunk 20 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp]: Reverted.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl (hunk 19 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl]: Added vectorization.

File: src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp (hunk 18 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/kernels/permute/permute_kernel_f_y_axes.cpp]: Changed for better naming.

File: src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp (hunk 22 above)
Comment (mhnap-lohika) [reply src/plugins/intel_gpu/tests/unit/test_cases/permute_gpu_test.cpp:2258]: For now, this kernel implementation doesn't support leftovers, so in the "Validate" method, we check only for aligned cases.

**Comment by mhnap-lohika** [comment]
> @mhnap-lohika Have you resolved all comments from Taylor?

Yes, it seems so.

File: src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl [hunk 23]
```diff
@@ -0,0 +1,200 @@
...
+        for (int i = 0; i < TILE_SIZE; ++i) {
+            const int x_idx = x_begin + i;
+            const int f = f_begin + j;
+            const int y_idx = y_begin + bf_local;
+#ifdef SUB_GROUP_SIZE
+            const uint output_offset = OUTPUT_GET_INDEX(b_idx, y_idx, f, x_idx) - get_sub_group_local_id();
+            DT_OUTPUT_BLOCK_WRITE(output, output_offset, transpose_buf[j][bf_local][i]);
+#else
+            const int output_idx = OUTPUT_GET_INDEX(b_idx, y_idx, f, x_idx);
+            output[output_idx] = transpose_buf[j][bf_local][i];
```
Comment (yeonbok) [inline src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl:109]: Can this pass benefit from this kernel? 
(No subgroup read, no vectorize). If not, how about get rid of this impl and prevent such case from the validation ?

Comment (mhnap-lohika) [reply src/plugins/intel_gpu/src/kernel_selector/cl_kernels/permute_f_y_axes.cl]: I also was thinking about it. It won't benefit, so I changed the "Validate" method for this case.

**Review by yeonbok — APPROVED** [review]
